### ISSUE TITLE
Add debug tooling for focus and IME connections

### ISSUE THREAD

**Issue by matthew-carroll** [flutter/flutter#118536]
As I'm working with user interfaces that have more complex focus and IME interactions, I'm finding it difficult to track associated changes.

I think it would be helpful to be able to visualize the focus path, from root, to the widget with primary focus. For example, all widgets in the focus path might be given a border and a tiny number in the corner that shows the tree distance from the root. The widget with primary focus could have a special color.

Also, I'd like to be able to track changes to the IME connection. For example, is an IME connection open right now? Ideally, I'd also like to know which widget owns that IME connection. That later part is more difficult, because there's no requirement that a widget own the IME connection. So maybe that requires an opt-in API. Nonetheless, when debugging issues at moments where one widget's IME connection closes and another is supposed to open, it would be helpful to have some quick toggle-able visual indicators through Flutter Inspector.

**Comment by matthew-carroll** [comment]
CC Flutter - @justinmc @gspencergoog 

CC - @knopp @miguelcmedeiros

**Comment by gspencergoog** [comment]
I agree that it would be nice to have a visual representation as a debugging aid.

Just to be sure you know about it: there is the `debugFocusChanges` boolean to turn on copious focus logging, including some of the things you are asking for, like printing the focus tree each time the focus changes. It's not as nice as having a visual representation on top of the widgets, but it logs the data.

**Comment by matthew-carroll** [comment]
Is that an option in Flutter Inspector? Or do you have to set that flag somewhere in the code?

**Comment by gspencergoog** [comment]
It is not an option in the Flutter Inspector. You need to set it somewhere.

**Comment by justinmc** [comment]
Is this something that would fall into your domain @kenzieschmoll?

I agree it sounds pretty useful to have these kinds of tools.

**Comment by loic-sharma** [comment]
It'd be useful to have a Flutter run key command to dump the focus tree. This would be similar to how you can `flutter run` and then press `w` to dump the widget hierarchy to console, or `t` to dump the rendering tree. The focus tree could be `f` or something.

**Comment by justinmc** [comment]
+1 to dumping the focus tree, that sounds pretty doable.

**Comment by gspencergoog** [comment]
There's already `debugDumpFocusTree` to call.

**Comment by loic-sharma** [comment]
Yup I've been using `debugDumpFocusTree` but the run key command has some advantages:

1. If you're outside of an IDE, there's no easy way to run `debugFocusTree` on demand
2. `debugDumpTree` in event callbacks results can be noisy
3. Run key commands are listed every time you `flutter run`, so they'd be a bit easier to discover than `debugFocusTree`

**Comment by gspencergoog** [comment]
Oh, I agree, I'm just pointing out what pressing the key should call.

**Comment by gspencergoog** [comment]
Sorry, that wasn't too clear. :-)

**Comment by loic-sharma** [comment]
Ah gotcha, I misread your comment. No worries :)

**Comment by loic-sharma** [comment]
In the next version of Flutter, `flutter run` now dumps the focus tree if the user presses `f`. See: https://github.com/flutter/flutter/pull/123473

There's also an open issue to add focus tooling in DevTools here: https://github.com/flutter/devtools/issues/5326
